src/libraries/System.Private.CoreLib/src/System/IntPtr.cs

```diff
@@ -34,136 +33,125 @@ namespace System
           IMinMaxValue<nint>,
           ISignedNumber<nint>
     {
-        private readonly unsafe void* _value; // Do not rename (binary serialization)
+        private readonly nint _value;
 
         [Intrinsic]
-        public static readonly IntPtr Zero;
+        public static readonly nint Zero;
 
         [NonVersionable]
-        public unsafe IntPtr(int value)
+        public IntPtr(int value)
         {
-            _value = (void*)value;
+            _value = value;
         }
 
         [NonVersionable]
-        public unsafe IntPtr(long value)
+        public IntPtr(long value)
         {
 #if TARGET_64BIT
-            _value = (void*)value;
+            _value = (nint)value;
 #else
-            _value = (void*)checked((int)value);
+            _value = checked((nint)value);
 #endif
         }
 
         [CLSCompliant(false)]
         [NonVersionable]
         public unsafe IntPtr(void* value)
         {
-            _value = value;
+            _value = (nint)value;
         }
 
-        private unsafe IntPtr(SerializationInfo info, StreamingContext context)
+        private IntPtr(SerializationInfo info, StreamingContext context)
         {
-            long l = info.GetInt64("value");
+            long value = info.GetInt64("value");
 
-            if (Size == 4 && (l > int.MaxValue || l < int.MinValue))
+#if TARGET_32BIT
+            if ((value > int.MaxValue) || (value < int.MinValue))
```

Comment:
A good example of where https://github.com/dotnet/runtime/issues/46259#issuecomment-748729659 and https://github.com/dotnet/runtime/issues/54439 would come in handy.

Reply:
I think I'd prefer the ifdef here anyways. This isn't exactly perf critical, but it is one of the "core types" in the BCL and we have the infrastructure to minimize the IL here.

Reply:
For libraries that aren't corelib and aren't getting built uniquely for each OS/Architecture, 100% agree